src/Application/UseCases/GetAllApproversCategories.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace ProfessionalWiki\PageApprovals\Application\UseCases;
+
+use ProfessionalWiki\PageApprovals\Adapters\UsersLookup;
+use ProfessionalWiki\PageApprovals\Adapters\DatabaseApproverRepository;
+use Wikimedia\Rdbms\IDatabase;
+use User;
+use MediaWiki\MediaWikiServices;
+use MediaWiki\Permissions\PermissionManager;
+
+class GetAllApproversCategories {
+
+	public const APPROVE_RIGHT = 'can-approve-pages';
+
+	public function __construct(
+		private readonly UsersLookup $usersLookup,
+		private readonly DatabaseApproverRepository $databaseApproverRepository
+	) {
+	}
+
+	/**
+	 * @return array<array{username: string, userId: int, categories: string[]}>
+	 */
+	public function getAllApproversCategories(): array {
+		$allUsers = $this->usersLookup->getAllUsers();
+		$approversWithCategories = [];
+
+		foreach ( $allUsers as $user ) {
+			if ( MediaWikiServices::getInstance()->getPermissionManager()->userHasRight(
+				$user,
+				self::APPROVE_RIGHT
+			) ) {
```

Comment:
There is room for a getAllApprovers function/service.

The overall implementation looks far from optimal. Getting ALL users, doing a permission check for each, and then another query per approver. The first two can likely be done in one DB query and maybe getting the categories can also be included there.

Reply:
A permission check is probably not needed. You can just see which users are present in the `approver_config` table.

Why not simply do a select on that table? Has all the info you need I suspect.

Reply:
Initially we will have a lot approvers that don't have any categories assigned yet or new approvers who are just added and don't have any categories. We need them, we have to do a permission check somewhere, I am not sure what's the cleanest way to do this.

Reply:
Moving most of the logic to the DB makes sense and it will also make testing a lot easier since we wouldn't have to mock so many classes